Handle UTF BOM when decoding text

radeusgd opened the pull request:
### Pull Request Description

- Improve BOM handling: detect and skip the BOM character, Default encoding that detects encoding based on BOM if present, warnings if unexpected BOM is encountered.
- Closes #9849 
- Windows-1252 fallback will be done as a separate PR as it has additional complexity. Tracked in ticket #10148.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

radeusgd committed:
Introduced Encoding.Default and switched read-like operations to use it as default/fallback

radeusgd committed:
tests for default encoding and UTF BOM

radeusgd committed:
tests for File read/write, what happens if invalid UTF but BOM encountered

radeusgd committed:
JSON test

radeusgd committed:
Delimited tests

radeusgd committed:
fix typos

radeusgd committed:
moving Encoding Utils

radeusgd committed:
updating Decoder

radeusgd committed:
update test

radeusgd committed:
DRY

radeusgd committed:
fix test

radeusgd committed:
more tests

radeusgd committed:
encoding detection based on BOM

radeusgd committed:
mark fallback tests as pending

radeusgd committed:
skip valid BOM in known UTF encodings

radeusgd committed:
refactor to have problem aggregator pattern

radeusgd committed:
javafmt

radeusgd committed:
detected BOM context in errors

radeusgd committed:
note

radeusgd committed:
report unexpected BOMs

radeusgd committed:
fixing tests

radeusgd committed:
check preconditions

radeusgd committed:
add some asserts, fix missing last partial character edge case

radeusgd committed:
fix test

radeusgd committed:
test edge case - currently parser has div by 0 ...

radeusgd committed:
fix most Delimited tests by allowing Default encoding in stream decoder

radeusgd committed:
javafmt

radeusgd committed:
test that error will include only a few positions

radeusgd committed:
fix some tests

radeusgd committed:
edge cases, asserts

radeusgd committed:
read several times to try to satisfy user request - this makes sure we will not return 0 from read, which was sometimes problematic for the usages

radeusgd committed:
more asserts

radeusgd committed:
annotation

radeusgd committed:
fix test

radeusgd committed:
adding annotation to on_problems to allow autoscoping

radeusgd committed:
changelog

radeusgd committed:
fix tests

AdRiley commented on std-bits/base/src/main/java/org/enso/base/encoding/ReportingStreamDecoder.java:
```diff
@@ -80,14 +98,28 @@ public ReportingStreamDecoder(InputStream stream, CharsetDecoder decoder) {
+   * Our read method tries to read as many characters as requested by the user.
    *
-   * <p>Used for reporting warnings.
+   * <p>It will never return 0 - it will either return a positive number of characters read, or -1
```
This comment seems to contradict line 120

radeusgd replied:
Ooops! I forgot about this edge case, I will update the doc 'will never return 0 unless user requested 0 bytes'. As the user setting `len == 0` should be the _only_ case where `0` would be returned - as guaranteed by asserts below.

radeusgd committed:
fix tests (2)

radeusgd committed:
ensure that we don't read the file all at once, but as needed

radeusgd committed:
fix comment

radeusgd committed:
fmt

radeusgd committed:
revert to Infer as it has different semantics from Default - it also considers metadata

jdunkerley commented on distribution/lib/Standard/AWS/0.0.0-dev/src/S3/S3_File.enso:
```diff
@@ -193,7 +193,7 @@ type S3_File
     @encoding Encoding.default_widget
     read_text : Encoding -> Problem_Behavior -> Text ! File_Error
-    read_text self (encoding=Encoding.utf_8) (on_problems=Problem_Behavior.Report_Warning) =
+    read_text self (encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```
```suggestion
    read_text self (encoding:Encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data.enso:
```diff
@@ -100,7 +100,7 @@ read path format=Auto_Detect (on_problems=Problem_Behavior.Report_Warning) = cas
 @encoding Encoding.default_widget
 read_text : (Text | File) -> Encoding -> Problem_Behavior -> Text
-read_text path (encoding=Encoding.utf_8) (on_problems=Problem_Behavior.Report_Warning) =
+read_text path (encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```
```suggestion
read_text path (encoding:Encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Encoding.enso:
```diff
@@ -48,19 +49,49 @@ type Encoding
        Arguments:
        - character_set: java.nio.charset name.
-    Value (character_set:Text)
+    Value (internal_character_set:Text)
```
```suggestion
    private Value (internal_character_set:Text)
```

radeusgd replied:
This cannot be done easily because we cannot have both private and public constructors mixed, and `Encoding.Default` is public.

We could make it private and create a 'factory' method `Encoding.default` I guess. Will amend in the next PR.

radeusgd replied:
Amended in this PR after all, for clarity.

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Encoding.enso:
```diff
@@ -48,19 +49,49 @@ type Encoding
+       Convert an Encoding to it's corresponding Java Charset, if applicable.
+       This method should be used in places not aware of special logic for the
+       Default encoding. In such places, usage of Default encoding will be forbidden.
+    to_java_charset self -> Charset ! Illegal_Argument =
+        self.to_java_charset_or_null.if_nothing <|
+            Error.throw (Illegal_Argument.Error "The Default encoding cannot be converted to a Java Charset. Please select a specific encoding for this operation.")
```
Would it be better to use UTF-8 and attach a warning?

Be good to get a feel where this is as we should probably think around how we workaround longer term.

radeusgd replied:
I don't think so, I did this deliberately.

My idea is that for Write operations we should use the `Encoding.utf_8` as the default _explicitly_ OR in places where read/write is both possible and `Encoding.Default` is used, our write logic should have logic to consciously replace `Encoding.Default` with something else in Write mode.

This is nicely demonstrated in commit 5656d6a30e6d331f47842e0963714fe2998e9853

If I had this method select UTF-8 automatically, I would not have noticed that I'm lacking special logic for Delimited Write - it would just write as UTF-8. But that would be incorrect - e.g. when appending to a file that upon read is detected to be UTF-16 because it has a BOM - in such cases we want to also append in UTF-16, so we needed special logic changes.

Thus IMO it is better to keep as is - because whenever this error occurs it should be one of 3 scenarios:
1) we mistakenly put `Encoding.Default` as a default value for a write-only operation,
2) we forgot to have special handling for `Encoding.Default` in a read/write operation,
3) the user explicitly selected `Encoding.Default` for a write operation.

In case of (3) the error makes sense because the user should just select an encoding explicitly.

In cases (1) and (2) it also means that we actually forgot to implement something and we should fix the code - otherwise we could get invalid behaviour and corrupted files - e.g. with mixed UTF-8 and UTF-16.

radeusgd replied:
Ah I think I somehow missed the 'attach a warning' part.

I still think the error will be more noticeable. But OTOH the warning will allow the user to proceed so maybe that is the right call indeed.

Amended.

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:
```diff
@@ -739,7 +739,7 @@ Text.is_whitespace self =
 @encoding Encoding.default_widget
 Text.bytes : Encoding -> Problem_Behavior -> Vector Integer
-Text.bytes self encoding on_problems=Problem_Behavior.Report_Warning =
+Text.bytes self encoding (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```
```suggestion
Text.bytes self encoding:Encoding (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```

We should probably default here as well.

radeusgd replied:
What would the default be? UTF-8?

radeusgd replied:
Added `UTF-8` default. Now this method by default has the same behaviour as `utf_8`. But I think it makes sense to keep both, to still have the `utf_8` shorthand if we want UTF-8 explicitly. The default for `bytes` is more oriented at the GUI/Component Browser usage where we want some default and it also happens to be the same.

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso:
```diff
@@ -756,15 +756,17 @@ Text.bytes self encoding on_problems=Problem_Behavior.Report_Warning =
-    result = Encoding_Utils.from_bytes bytes (encoding . to_java_charset) : ResultWithWarnings
-    if result.warnings.is_nothing then result.result else
-        on_problems.attach_problems_after result.result [Encoding_Error.Error result.warnings]
+Text.from_bytes bytes encoding (on_problems : Problem_Behavior = Problem_Behavior.Report_Error) =
```
```suggestion
Text.from_bytes bytes encoding:Encoding=Encoding.Default (on_problems : Problem_Behavior = Problem_Behavior.Report_Error) =
```

Feels like we should default now.

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Enso_File.enso:
```diff
@@ -290,7 +290,7 @@ type Enso_File
     @encoding Encoding.default_widget
     read_text : Encoding -> Problem_Behavior -> Text ! File_Error
-    read_text self (encoding=Encoding.utf_8) (on_problems=Problem_Behavior.Report_Warning) =
+    read_text self (encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```
```suggestion
    read_text self (encoding:Encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso:
```diff
@@ -333,7 +333,7 @@ type File
     @encoding Encoding.default_widget
     read_text : Encoding -> Problem_Behavior -> Text ! File_Error
-    read_text self (encoding=Encoding.utf_8) (on_problems=Problem_Behavior.Report_Warning) =
+    read_text self (encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```
```suggestion
    read_text self (encoding:Encoding=Encoding.Default) (on_problems : Problem_Behavior = Problem_Behavior.Report_Warning) =
```

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9849-auto-utf

radeusgd committed:
fix Delimited Writer:
Encoding.Default is treated as UTF-8 for new files
when appending, Encoding.Default will try to detect the encoding

jdunkerley reviewed (approved):
Some little things for next PR...

radeusgd committed:
add tests that write uses the encoding like read would

radeusgd committed:
make Encoding constructor private

radeusgd committed:
CR: adding type annotations

radeusgd committed:
CR: change error to warning

radeusgd committed:
CR: add defaults

radeusgd committed:
Merge branch 'develop' into wip/radeusgd/9849-auto-utf

radeusgd committed:
fixing tests

radeusgd committed:
Merge branch 'refs/heads/develop' into wip/radeusgd/9849-auto-utf

radeusgd committed:
fix NI test
